staging/src/k8s.io/apiserver/pkg/cel/environment/base.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package environment
+
+import (
+	"sync/atomic"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/ext"
+
+	"k8s.io/apimachinery/pkg/util/version"
+	celconfig "k8s.io/apiserver/pkg/apis/cel"
+	"k8s.io/apiserver/pkg/cel/library"
+)
+
+// DefaultCompatibilityVersion returns a default compatibility version for use with EnvSet
+// that guarantees N-1 rollbacks are safe.
+//
+// This default will be set to no more than N-1 the current Kubernetes major.minor version.
+//
+// Note that a default version number less than N-1 indicates a wider range of version
+// compatibility than strictly required for rollback. A wide range of compatibility is
+// desirable because it means that CEL expressions are portable across a wider range
+// of Kubernetes versions.
+func DefaultCompatibilityVersion() *version.Version {
+	return defaultCompatibilityVersion
+}
+
+var defaultCompatibilityVersion = version.MajorMinor(1, 27)
+
+var baseOpts = []VersionedOptions{
```

Comment:
> We will never allow func/lib/variables be back ported
Correct. This would result in clearly user visible feature differences between point releases (e.g. 1.27.1 and 1.272) which is something we don't want.

> deletion/deprecation case right?
We're don't have a viable policy for deletion/deprecation, and given that we'll be writing expressions to Kubernetes resources, the backward compatibility requirements are really important here.